packages/start/config/server-runtime.jsx

```diff
@@ -13,55 +13,100 @@ import {
 } from 'seroval-plugins/web';
 import { createIslandReference } from "../server/islands";
 
-async function deserializeStream(id, response) {
-  if (!response.body) {
-    throw new Error("missing body");
+class ChunkReader {
+  constructor(stream) {
+    this.reader = stream.getReader();
+    this.buffer = '';
+    this.done = false;
   }
-  const reader = response.body.getReader();
-
-  async function pop() {
-    const result = await reader.read();
-    if (!result.done) {
-      const serialized = new TextDecoder().decode(result.value);
-      const splits = serialized.split("\n");
-      for (const split of splits) {
-        if (split !== "") {
-          deserialize(split);
-        }
-      }
-      await pop();
+
+  async readChunk() {
+    // if there's no chunk, read again
+    const chunk = await this.reader.read();
+    if (!chunk.done) {
+      // repopulate the buffer
+      this.buffer += new TextDecoder().decode(chunk.value);
+    } else {
+      this.done = true;
     }
   }
 
-  let serialized = ""
-  while (true) {
-    const line = await reader.read();
-    if (line.done) break;
-    serialized += new TextDecoder().decode(line.value);
+  async nextValue() {
+    // Check if the buffer is empty
+    if (this.buffer === '') {
+      // if we are already one...
+      if (this.done) {
+        return {
+          done: true,
+          value: undefined,
+        };
+      }
+      // Otherwise, read a new chunk
+      await this.readChunk();
+    }
+    // Get the first valid seroval chunk
+    const [first, ...rest] = this.buffer.split('\n');
+    // Deserialize the seroval chunk
+    const result = {
+      done: false,
+      value: deserialize(first),
+    };
+    // if it succeeds, remove the first valid chunk
+    // from the buffer
+    this.buffer = rest.join('\n');
+    return result;
   }
-  let pending = true;
-  let revived;
-  const splits = serialized.split("\n");
-  for (const split of splits) {
-    if (split !== "") {
-      const current = deserialize(split);
-      if (pending) {
-        revived = current;
-        pending = false;
+
+  async next() {
+    try {
+      // Attempt to read a valid seroval chunk
+      return await this.nextValue();
+    } catch (error) {
+      // If it happens that there's an error again
+      // and we are done reading the buffer
+      // then the whole stream is invalid.
+      if (this.done) {
+        throw new Error('Malformed server function stream.');
       }
+      // Since it's invalid (some syntax-related issue)
+      // we read a new chunk, and hope there's a valid
```

Comment:
I will test this tomorrow, but do we think this will fix #1245 where a single large JS expression is split across two chunks? AFAIK we need to concatenate the two partial chunks so that the entire chunk is a valid expression but I think ATM this code would drop both chunks?

Reply:
I've already tested this with a gigantic chunk (as seen with the Test app). The idea is to keep reading until there's a valid seroval (JS) chunk.

Reply:
Can you help explain to me? My understanding is that on an invalid chunk https://github.com/solidjs/solid-start/pull/1245/files#diff-d5889473d46bb02f0141e0e1dfa59c3c5c118a8e660eb941117bfc25fd3d1234R52 will throw and `first` will lose all references and be GCed.

Reply:
Basically, when we call `deserialize` we are also trying to know if the `first` chunk is a valid seroval chunk. Once it deserializes successfully, we remove the `first` chunk from the `buffer`.

If `deserialize` fails then we move to the `catch` block. There we read the next Response chunk, and then attempt to deserialize again. Take note that the `buffer` only shrinks for every successful deserialize call.

We repeat this until a valid seroval chunk has been encountered and/or the response is done and the `buffer` is empty.

Reply:
Thanks for helping me out, I get it now. A reader of L52 must be aware of exception handling on L68 in order to reason about correct behavior

Reply:
TBF that thrown Error is unlikely to happen, I guess.